File: src/cli/services/commands/diff.ts
// SPDX-License-Identifier: Apache-2.0

import * as fs from "fs";
import * as path from "path";
import {
    normalizeBoardModel,
    parseBoardModel,
} from "../../../configurator/service";
import { ProjectSettings } from "../../../project/models";
import { resolveProjectContext } from "../../../project/service";
import {
    CLI_EXIT_CODE,
    CliExecutionInput,
    CliExecutionResult,
    CliGlobalFlags,
} from "../../models";
import { createEnvelope, createFailureResult } from "../envelope";

type DiffKind = "added" | "removed" | "changed";

interface DiffEntry {
  path: string;
  kind: DiffKind;
  before: unknown;
  after: unknown;
}

interface DiffCommandData {
  schemaVersion: "1";
  boardYamlPath: string;
  unchanged: boolean;
  changeCount: number;
  changes: DiffEntry[];
}

export function runDiffCommand(
  flags: CliGlobalFlags,
  input: Omit<CliExecutionInput, "argv">,
): CliExecutionResult<DiffCommandData> {
  try {
    const workspaceRoot = path.resolve(input.cwd, flags.projectPath ?? ".");
    const settings: ProjectSettings = {
      sdkPath: flags.sdkRoot ?? "",
      pythonPath: "",
      boardYamlPath: flags.boardYamlPath ?? "board.yaml",
      westCwd: "",
    };
    const context = resolveProjectContext(
      {
        workspaceFolders: [workspaceRoot],
        settings,
        platform: input.platform,
      },
      input.pathExists,
    );

    if (!context.boardYamlPath || !input.pathExists(context.boardYamlPath)) {
      return createFailureResult(
        "diff",
        flags.format,
        CLI_EXIT_CODE.validationFailure,
        ["diff: board.yaml path is unresolved or missing."],
        [
          {
            code: "diff.board-yaml-missing",
            severity: "error",
            message:
              "board.yaml path could not be resolved or the file does not exist.",
          },
        ],
        {
          schemaVersion: "1",
          boardYamlPath: context.boardYamlPath ?? "",
          unchanged: false,
          changeCount: 0,
          changes: [],
        },
      );
    }

    const boardText = fs.readFileSync(context.boardYamlPath, "utf8");
    const boardModel = parseBoardModel(boardText);
    const normalized = normalizeBoardModel(boardModel);

    const changes = collectDiffEntries(boardModel, normalized);
    const textLines =
      flags.format === "json"
        ? []
        : formatDiffTextLines(changes, context.boardYamlPath, flags.quiet);

    return {
      format: flags.format,
      exitCode: CLI_EXIT_CODE.success,
      textLines,
      envelope: createEnvelope(
        "diff",
        {
          root: context.workspaceRoot,
          boardYaml: context.boardYamlPath,
        },
        {
          schemaVersion: "1",
          boardYamlPath: context.boardYamlPath,
          unchanged: changes.length === 0,
          changeCount: changes.length,
          changes,
        },
        [],
        CLI_EXIT_CODE.success,
      ),
    };
  } catch (error) {
    const message =
      error instanceof Error ? error.message : "Unexpected CLI diff failure.";
    return createFailureResult(
      "diff",
      flags.format,
      CLI_EXIT_CODE.internalFailure,
      ["diff: internal failure", message],
      [
        {
          code: "diff.internal-failure",
          severity: "error",
          message,
        },
      ],
      {
        schemaVersion: "1",
        boardYamlPath: "",
        unchanged: false,
        changeCount: 0,
        changes: [],
      },
    );
  }
}

function collectDiffEntries(before: unknown, after: unknown): DiffEntry[] {
  const output: DiffEntry[] = [];
  collectRecursiveDiff(before, after, "", output);
  return output.sort((left, right) => left.path.localeCompare(right.path));
}

function collectRecursiveDiff(
  before: unknown,
  after: unknown,
  currentPath: string,
  output: DiffEntry[],
): void {
  if (deepEqual(before, after)) {
    return;
  }

  if (isRecord(before) && isRecord(after)) {
    const keys = new Set<string>([
      ...Object.keys(before),
      ...Object.keys(after),
    ]);

    for (const key of [...keys].sort()) {
      const nextPath = currentPath ? `${currentPath}.${key}` : key;
      const hasBefore = Object.prototype.hasOwnProperty.call(before, key);
      const hasAfter = Object.prototype.hasOwnProperty.call(after, key);

      if (!hasBefore) {
        output.push({
          path: nextPath,
          kind: "added",
          before: undefined,
          after: after[key],
        });
        continue;
      }

      if (!hasAfter) {
        output.push({
          path: nextPath,
          kind: "removed",
          before: before[key],
          after: undefined,
        });
        continue;
      }

      collectRecursiveDiff(before[key], after[key], nextPath, output);
    }
    return;
  }

  if (Array.isArray(before) && Array.isArray(after)) {
    const maxLength = Math.max(before.length, after.length);
    for (let index = 0; index < maxLength; index += 1) {
      const nextPath = `${currentPath}[${index}]`;

      if (index >= before.length) {
        output.push({
          path: nextPath,
          kind: "added",
          before: undefined,
          after: after[index],
        });
        continue;
      }

      if (index >= after.length) {
        output.push({
          path: nextPath,
          kind: "removed",
          before: before[index],
          after: undefined,
        });
        continue;
      }

      collectRecursiveDiff(before[index], after[index], nextPath, output);
    }
    return;
  }

  output.push({
    path: currentPath || "<root>",
    kind: "changed",
    before,
    after,
  });
}

function deepEqual(left: unknown, right: unknown): boolean {
  if (left === right) {
    return true;
  }

  if (Array.isArray(left) && Array.isArray(right)) {
    if (left.length !== right.length) {
      return false;
    }

    for (let index = 0; index < left.length; index += 1) {
      if (!deepEqual(left[index], right[index])) {
        return false;
      }
    }

    return true;
  }

  if (isRecord(left) && isRecord(right)) {
    const leftKeys = Object.keys(left).sort();
    const rightKeys = Object.keys(right).sort();

    if (!deepEqual(leftKeys, rightKeys)) {
      return false;
    }

    for (const key of leftKeys) {
      if (!deepEqual(left[key], right[key])) {
        return false;
      }
    }

    return true;
  }

  return false;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function formatDiffTextLines(
  changes: readonly DiffEntry[],
  boardYamlPath: string,
  quiet: boolean,
): string[] {
  if (changes.length === 0) {
    return ["diff: no effective-config differences detected."];
  }

  const lines = [`diff: ${changes.length} differences in ${boardYamlPath}`];

  if (!quiet) {
    for (const change of changes) {
      lines.push(
        `${change.kind.toUpperCase()} ${change.path}: ${formatValue(change.before)} -> ${formatValue(change.after)}`,
      );
    }
  }

  return lines;
}

function formatValue(value: unknown): string {
  if (value === undefined) {
    return "<undefined>";
  }

  if (typeof value === "string") {
    return JSON.stringify(value);
  }

  const raw = JSON.stringify(value);
  if (!raw) {
    return String(value);
  }

  if (raw.length > 120) {
    return `${raw.slice(0, 117)}...`;
  }

  return raw;
}

File: cli-rs/crates/alp-cli/src/commands/diff.rs
// SPDX-License-Identifier: Apache-2.0
//! `alp diff` — show how `normalize_board_model` changes the parsed board.yaml.
//!
//! Mirrors TS `runDiffCommand`: parse board.yaml, normalize it, and diff the
//! two as JSON trees (added/removed/changed by path). Null object entries are
//! pruned first so the diff matches the TS model's sparse (undefined-omitting)
//! shape.

use std::path::Path;

use alp_core::{
    DiffEntry, DiffKind, collect_diff_entries, normalize_board_model, parse_board_model,
    prune_nulls,
};
use serde_json::Value;

use super::CommandRun;
use crate::cli::GlobalArgs;
use crate::envelope::{Envelope, Issue, Project};
use crate::exit::ExitCode;
use crate::util::resolve_cli_project_context;

#[derive(serde::Serialize)]
struct DiffData {
    #[serde(rename = "schemaVersion")]
    schema_version: String,
    #[serde(rename = "boardYamlPath")]
    board_yaml_path: String,
    unchanged: bool,
    #[serde(rename = "changeCount")]
    change_count: usize,
    changes: Vec<DiffEntry>,
}

pub fn run(g: &GlobalArgs) -> CommandRun {
    let context = resolve_cli_project_context(g);
    let project = Project {
        root: context.workspace_root.clone(),
        board_yaml: context.board_yaml_path.clone(),
    };

    let board_path = match &context.board_yaml_path {
        Some(path) if Path::new(path).exists() => path.clone(),
        _ => {
            return failure(
                g,
                project,
                ExitCode::ValidationFailure,
                "board-yaml-missing",
                "board.yaml path could not be resolved or the file does not exist.",
                context.board_yaml_path.clone().unwrap_or_default(),
                vec!["diff: board.yaml path is unresolved or missing.".to_string()],
            );
        }
    };

    let text = match std::fs::read_to_string(&board_path) {
        Ok(t) => t,
        Err(e) => {
            return failure(
                g,
                project,
                ExitCode::InternalFailure,
                "internal-failure",
                &format!("could not read board.yaml: {e}"),
                board_path.clone(),
                vec!["diff: internal failure".to_string(), e.to_string()],
            );
        }
    };

    let model = match parse_board_model(&text) {
        Ok(m) => m,
        Err(e) => {
            return failure(
                g,
                project,
                ExitCode::InternalFailure,
                "internal-failure",
                &e.to_string(),
                board_path.clone(),
                vec!["diff: internal failure".to_string(), e.to_string()],
            );
        }
    };

    let before = prune_nulls(serde_json::to_value(&model).expect("board model is serializable"));
    let normalized = normalize_board_model(model);
    let after =
        prune_nulls(serde_json::to_value(&normalized).expect("board model is serializable"));
    let changes = collect_diff_entries(&before, &after);

    let data = DiffData {
        schema_version: "1".to_string(),
        board_yaml_path: board_path.clone(),
        unchanged: changes.is_empty(),
        change_count: changes.len(),
        changes,
    };

    let text_lines = if g.is_json() {
        Vec::new()
    } else {
        format_diff_text(&data, g, &board_path)
    };
    let json = g.is_json().then(|| {
        Envelope::new("diff", project, data, Vec::new(), ExitCode::Success.code()).to_json()
    });

    CommandRun {
        exit: ExitCode::Success,
        text: text_lines,
        json,
    }
}

#[allow(clippy::too_many_arguments)]
fn failure(
    g: &GlobalArgs,
    project: Project,
    exit: ExitCode,
    code: &str,
    message: &str,
    board_yaml_path: String,
    text_lines: Vec<String>,
) -> CommandRun {
    let issues = vec![Issue {
        code: format!("diff.{code}"),
        severity: "error".to_string(),
        message: message.to_string(),
    }];
    let data = DiffData {
        schema_version: "1".to_string(),
        board_yaml_path,
        unchanged: false,
        change_count: 0,
        changes: Vec::new(),
    };
    let text = if g.is_json() { Vec::new() } else { text_lines };
    let json = g
        .is_json()
        .then(|| Envelope::new("diff", project, data, issues, exit.code()).to_json());

    CommandRun { exit, text, json }
}

fn format_diff_text(data: &DiffData, g: &GlobalArgs, board_path: &str) -> Vec<String> {
    if data.changes.is_empty() {
        return vec!["diff: no effective-config differences detected.".to_string()];
    }

    let mut lines = vec![format!(
        "diff: {} differences in {}",
        data.changes.len(),
        board_path
    )];
    if !g.quiet {
        for change in &data.changes {
            lines.push(format!(
                "{} {}: {} -> {}",
                kind_label(change.kind),
                change.path,
                format_value(&change.before),
                format_value(&change.after),
            ));
        }
    }
    lines
}

fn kind_label(kind: DiffKind) -> &'static str {
    match kind {
        DiffKind::Added => "ADDED",
        DiffKind::Removed => "REMOVED",
        DiffKind::Changed => "CHANGED",
    }
}

fn format_value(value: &Option<Value>) -> String {
    match value {
        None => "<undefined>".to_string(),
        Some(Value::String(s)) => serde_json::to_string(s).unwrap_or_else(|_| s.clone()),
        Some(val) => {
            let raw = serde_json::to_string(val).unwrap_or_else(|_| val.to_string());
            if raw.chars().count() > 120 {
                let truncated: String = raw.chars().take(117).collect();
                format!("{truncated}...")
            } else {
                raw
            }
        }
    }
}
